sql/src/main/java/io/crate/metadata/Schemas.java

```diff
@@ -21,25 +21,244 @@
+        if (!(tableInfo instanceof DocTableInfo)) {
+            throw new UnsupportedOperationException(String.format(Locale.ENGLISH,
+                "The table %s is read-only. Write, Drop or Alter operations are not supported", tableInfo.ident()));
+        }
+        DocTableInfo docTableInfo = (DocTableInfo) tableInfo;
+        if (docTableInfo.isAlias() && !docTableInfo.isPartitioned()) {
+            throw new UnsupportedOperationException(String.format(Locale.ENGLISH,
+                "%s is an alias. Write, Drop or Alter operations are not supported", tableInfo.ident()));
+        }
+        return docTableInfo;
+    }
+
+    /**
+     * @param ident the table ident to get a TableInfo for
+     * @return an instance of TableInfo for the given ident, guaranteed to be not null
+     * @throws io.crate.exceptions.SchemaUnknownException if schema given in <code>ident</code>
+     *                                                    does not exist
+     * @throws io.crate.exceptions.TableUnknownException  if table given in <code>ident</code> does
+     *                                                    not exist in the given schema
+     */
+    public TableInfo getTableInfo(TableIdent ident) {
+        SchemaInfo schemaInfo = getSchemaInfo(ident);
+        TableInfo info;
+        info = schemaInfo.getTableInfo(ident.name());
+        if (info == null) {
+            throw new TableUnknownException(ident);
+        }
+        return info;
+    }
+
+    private SchemaInfo getSchemaInfo(TableIdent ident) {
+        String schemaName = firstNonNull(ident.schema(), DEFAULT_SCHEMA_NAME);
+        SchemaInfo schemaInfo = schemas.get(schemaName);
+        if (schemaInfo == null) {
+            throw new SchemaUnknownException(schemaName);
+        }
+        return schemaInfo;
+    }
+
+    @Nonnull
+    public Iterator<SchemaInfo> iterator() {
+        return schemas.values().iterator();
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (!event.metaDataChanged()) {
+            return;
+        }
+
+        Set<String> newCurrentSchemas = getNewCurrentSchemas(event.state().metaData());
+        synchronized (schemas) {
+            Sets.SetView<String> nonBuiltInSchemas = Sets.difference(schemas.keySet(), builtInSchemas.keySet());
+            Set<String> deleted = Sets.difference(nonBuiltInSchemas, newCurrentSchemas).immutableCopy();
+            Set<String> added = Sets.difference(newCurrentSchemas, schemas.keySet()).immutableCopy();
+
+            for (String deletedSchema : deleted) {
+                try {
+                    schemas.remove(deletedSchema).close();
+                } catch (Exception e) {
+                    LOGGER.error(e.getMessage(), e);
+                }
+            }
+
+            // trigger changed event on all existing schemas
+            for (SchemaInfo schemaInfo : this) {
+                schemaInfo.clusterChanged(event);
```

Comment:
maybe rename that clusterChanged method to something like update, in order to not confuse it with the listener method.